### ISSUE TITLE
File run causes cov report messed up

### ISSUE THREAD

**Issue by quolpr** [fredrikaverpil/neotest-golang#23]
I am not sure how test is running for current test file (with `require("neotest").run.run(vim.fn.expand("%"))`), but the coverage report output is actually for one of the tests cases, instead of summing up of all tests.

**Comment by fredrikaverpil** [comment]
Yes, this is (by me) known, actually. It's because the tests are run one by one, even when you run all tests in a file (and when running the entire test suite).

Because of this, you'll always only see the coverage of the last executed test at the end of the execution.

The reason why it works like this is it was the fastest and simplest way to get everything working.

I've been wanting to optimize "run test file" and "run test suite" so that the test results are parsed/collected and the neotest node tree is updated accordingly. This brings a number of benefits such as also being a lot faster.

Unfortunately, there's no workaround for now.

**Comment by quolpr** [comment]
@fredrikaverpil yeah, make sense 👍 One solution is spinning in my mind, what if we will be collecting all test names from file, and the just run them with usual grep, like:

`go test -v magnit/mag_collect_service/internal/service/order/issuer -run "^(TestIssuer_NotFound|TestIssuer_WrongState|TestIssuer_IssueCourier_WithCode)" -coverprofile ./coverage.out`

Will it be easy to implement?

**Comment by fredrikaverpil** [comment]
@quolpr FYI I made some great progress tonight on all this in #24 but it still requires a great deal of work and then a great deal of testing during daily work... so to make sure I catch all bugs and gotchas.

**Comment by quolpr** [comment]
@fredrikaverpil 
Wow, such a detailed comment. Thank you for the clarification! Actually, after everything you described, it seems the coverage report might not be a necessary feature, considering the amount of effort required to introduce it. Additionally, it could potentially be buggy.

By the way, don't rush yourself—it's open source! 😅

**Comment by quolpr** [comment]
Also, it's sad that neotest is somehow buggy and we can't use just JSON output. That would solve a lot of problems 🤔 But maintainers don't look to be active for now

**Comment by fredrikaverpil** [comment]
> it seems the coverage report might not be a necessary feature, considering the amount of effort required to introduce it.

Coverage is always going to be yanky unless you know exactly what you want to measure, and understand which command is the appropriate one to measure that. In this case, if you run a single test case, you'll get coverage for that. But if you run all tests of a file, or all test of a folder, my goal is you should get coverage for that. I view the current behavior of always running tests on a test-by-test case as an "MVP", if you will. It was the fastest way to get something up and working while also learning the Neotest framework.

Proper coverage support by itself might not be a necessary feature, like you say. But the test execution is also extremely inefficient right now. So this is something I want to address, and I believe I've found a great way forward. But I'm starting with "run all tests in a folder (and sub-folders)" as that is a little easier, and can be built upon for "run all tests of a file".

> neotest is somehow buggy and we can't use just JSON output.

I guess you are referring to how you get the stdout/stderr output mixed into parsed test results, as I reported in https://github.com/nvim-neotest/neotest/issues/391?

I've found a pretty good workaround though, implemented in the adapter right now, which circumvents that. Basically, the raw output is parsed (as JSON) and afterwards, the file on disk containing the original output is erased, as if there was no output from `go test`. This means that there is nothing for Neotest to "leak" into its test output window. I think there are no side-effects of this, actually. But I agree, it would be better to get this fixed upstream.

**Comment by fredrikaverpil** [comment]
@quolpr in case you want to help out test what I have in #24, just set the `branch` for your package manager. 😄 

It's not exactly what you asked for, as that branch enables running a folder of tests (although it's more of a go package). But it would still be valuable with your input/feedback on how e.g. coverage calculation works for you with this.

**Comment by quolpr** [comment]
@fredrikaverpil wow, thank you for all the work, but I decided to write my own plugin https://github.com/quolpr/quicktest.nvim because I find neotest to be too bloated for me 🤔

**Comment by fredrikaverpil** [comment]
Nice work @quolpr ! 👏 🎉 

No worries, I wanted to look into this anyway but your input prompted me to look into this sooner than later. Feel free to close this issue if you no longer care to track this. I've still got this on my radar for my adapter.

**Comment by fredrikaverpil** [comment]
Just a quick headsup I'm about to merge #92 which fixes this. 😄
